src/frontend/contexts/ProjectContext.tsx

```diff
@@ -0,0 +1,89 @@
+import * as React from 'react';
+import {type MapeoClientApi} from '@mapeo/ipc';
+import {useApi} from './ApiContext';
+import {usePersistedProjectId} from '../hooks/persistedState/usePersistedProjectId';
+
+type MapeoProject = Awaited<ReturnType<MapeoClientApi['getProject']>>;
+
+type ActiveProjectContext = {
+  project: MapeoProject | undefined;
+  switchProject(projectId: string): void;
+};
+
+export const ActiveProjectContext = React.createContext<
+  ActiveProjectContext | undefined
+>(undefined);
+
+export const ActiveProjectProvider = ({
+  children,
+}: React.PropsWithChildren<{}>) => {
+  const mapeoApi = useApi();
+  const [activeProject, setActiveProject] = React.useState<
+    MapeoProject | undefined
+  >();
+  const activeProjectId = usePersistedProjectId(store => store.projectId);
+  const setActiveProjectId = usePersistedProjectId(store => store.setProjectId);
+
+  React.useEffect(() => {
+    if (!activeProjectId) return;
+    let cancelled = false;
+    mapeoApi
+      .getProject(activeProjectId)
+      .then(project => {
+        if (cancelled) return;
+        setActiveProject(project);
+      })
+      .catch(error => {
+        if (cancelled) return;
+
+        setActiveProjectId(undefined);
+      });
+    return () => {
+      cancelled = true;
+    };
+  }, [mapeoApi, activeProjectId, setActiveProjectId, setActiveProject]);
+
+  // If no active projectId is set, then either get the first existing project,
+  // or if no projects yet exist, create a project
+  React.useEffect(() => {
+    if (activeProjectId) return;
+    let cancelled = false;
+    getExistingProjectIdOrCreate(mapeoApi)
+      .then(projectId => {
+        if (cancelled) return;
+        setActiveProjectId(projectId);
+      })
+      .catch(error => {
+        // TODO: Store in state and render error
+        console.error(error);
+      });
+    return () => {
+      cancelled = true;
+    };
+  }, [activeProjectId, setActiveProjectId]);
+
+  const contextValue = React.useMemo<ActiveProjectContext>(() => {
+    return {
+      project: activeProject,
+      switchProject(projectId: string) {
+        setActiveProjectId(projectId);
+      },
+    };
+  }, [activeProject, setActiveProjectId]);
+
+  return (
+    <ActiveProjectContext.Provider value={contextValue}>
+      {children}
+    </ActiveProjectContext.Provider>
+  );
+};
+
+// If we don't have an activeProjectId set, then we use either the first project
+// returned by mapeo.listProjects() or we create a new default project
+async function getExistingProjectIdOrCreate(mapeoApi: MapeoClientApi) {
+  const projects = await mapeoApi.listProjects();
+  // We shouldn't normally get here, except when somehow we end up with an
+  // invalid active project id
+  if (projects[0]) return projects[0].projectId;
+  return mapeoApi.createProject();
+}
```

Comment:
@gmaclennan This might cause a problem if the user does not have a way of switching a project (which is not possible for the mvp). 

If the user joins a new project, they will have 2 projects. The front end will save the id of the new project to persisted state. But if the persisted state cache gets cleared, this functions behaviour will put the user to the first project, which is presumably the default project. Since there is not a way of switching projects in the front end, the user will not be able to access the other project. 

Down the line, that will be ok because they will be able switch between projects. But for the MVP, we could possibly kick people out of their projects by accident.

Reply:
Is there a reason that persisted state would be cleared? Is it less likely to happen on the backend than the front end?

Reply:
Good point, Im actually just assuming that is easier to clear the cache in the front end. A user can clear data cache and data from an app in android settings. That will delete persisted storage in the front end. I don't actually know the architecture of how information is persisted with the backend, so I don't know how those actions would effect it.

I'm happy to keep your architecture as well for simplicity. This is an edge case that won't last for long once we have projects.

Reply:
By default mmkv [stores data in files storage](https://github.com/Tencent/MMKV/wiki/android_advance#customize-location) (not cache storage). It would be deleted if the user deleted the app data, but not if they deleted the app cache. However the backend works the same way - there is no data that can persist across the user deleting app data. Therefore I think it's fine to persist this in the frontend - there is no extra security against deletion that comes with storing in the backend. We just need to be careful that the app does not delete this piece of persisted data

Reply:
wasn't aware that mmkv doesn't use cache storage, which was my main concern around keeping the active project id persisted in the frontend. given that, think gregor's assertion about it living there make sense 👍